### PR TITLE
[Number Series Copilot] Bug Fixes in Intent Understanding and Field Validation

### LINKED ISSUE #2106: [Bug]: [Number Series Copilot] Bug fixes
### **Bugs**

1. **Bug 1: Incorrect Intent Mapping for "Generate" Command**

   - **Problem**: When users intended to create a new Number Series using the word “generate,” the system often mistakenly interpreted the command as a request to prepare Number Series for the next period (e.g., the next fiscal year). This is a misinterpretation of user intent.
   - **Expected Behavior**: The LLM should correctly distinguish between creating a new Number Series (using terms like "create" or "generate") and preparing a Number Series for the next year/period (using more specific time-related phrases).
   - **Proposed Fix**: Update the tools definition and prompt to clarify the distinction between creating new Number Series and generating Number Series for the next period. Ensure that the system identifies the intent based on contextual clues rather than assuming a time-based action when the word "generate" is used.

2. **Bug 2: Incorrect Field Check for Existing Number Series**

   - **Problem**: When a user generated Number Series and some already existed in the database, the LLM incorrectly performed checks on all fields (e.g., warning number, increment number, ending number), which should not be the case for pre-existing series. Only new Number Series should undergo a full check.
   - **Expected Behavior**: The system should skip validation of non-mandatory fields (such as warning number and increment number) for existing Number Series and only perform these checks on newly created series.
   - **Proposed Fix**: Modify the behavior to exclude pre-existing Number Series from the field validation checks. Ensure that only mandatory fields (such as Number Series code, description, and series) are validated for new series.

3. **Bug 3: Blank Fields Causing Validation Errors**

   - **Problem**: Sometimes the LLM generated Number Series without values for optional fields like the warning number or increment number. If the user's environment had blank fields in an existing Number Series, the system would replicate this and then fail validation when mandatory fields were checked.
   - **Expected Behavior**: The system should only validate mandatory fields, such as the Number Series code and description, and allow optional fields to remain blank if necessary. This would prevent errors from being thrown unnecessarily.
   - **Proposed Fix**: Refine the validation logic so that only mandatory fields are checked. Optional fields, such as warning number and increment number, should not cause validation failures if left blank.

### **Steps to Implement:**

1. **Bug Fixes**:
    - Refactor the LLM prompt and tools definition to improve the understanding of user intent between creating and preparing Number Series for different periods.
    - Update the logic to exclude field checks for existing Number Series and restrict checks to mandatory fields.

### **Acceptance Criteria**:
- Users can create Number Series without mistakenly triggering next-period Number Series generation when using the word "generate."
- Existing Number Series are excluded from unnecessary checks, and only mandatory fields are validated for new series.

### Additional context

_No response_

### I will provide a fix for a bug

- [X] I will provide a fix for a bug

### PR DESCRIPTION
**Pull request by DmitryKatson**
#### Summary 

This PR addresses three critical bugs in the Number Series Copilot:

1. **Bug 1: Incorrect Intent Mapping for "Generate" Command**
   - The system was misinterpreting the "generate" command, often preparing Number Series for the next period instead of creating a new one. The fix involves updating the tools definition and prompt to clarify the distinction between creating new Number Series and generating Number Series for the next period.

2. **Bug 2: Incorrect Field Check for Existing Number Series**
   - The system was incorrectly performing checks on all fields for existing Number Series. The fix modifies the behavior to exclude pre-existing Number Series from the field validation checks, ensuring that only mandatory fields are validated for AI generated series.

3. **Bug 3: Blank Fields Causing Validation Errors**
   - The system was generating Number Series without values for optional fields, causing validation errors. The fix refines the validation logic so that only mandatory fields are checked, allowing optional fields to remain blank if necessary.

**Steps to Implement:**

1. Refactor the LLM prompt and tools definition to improve the understanding of user intent between creating and preparing Number Series for different periods.
2. Update the logic to exclude field checks for existing Number Series and restrict checks to mandatory fields.

**Acceptance Criteria:**

- Users can create Number Series without mistakenly triggering next-period Number Series generation when using the word "generate."
- Existing Number Series are excluded from unnecessary checks, and only mandatory fields are validated for new series.

These fixes aim to enhance the accuracy and reliability of the Number Series Copilot, ensuring it correctly interprets user commands and validates the necessary fields.

#### Work Item(s) 

Fixes #2106 

Fixes [AB#550892](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/550892), [AB#544073](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/544073)

### CODE REVIEW

**Comment by DmitryKatson** [comment]
@JesperSchulz you should update:
- ToolsSelectionPrompt.txt
- Tool3Definition.txt
BUT Important! These should have new version starting from v25.1 . For the v25.0 they should stay current, as they aligned with the function in AL that changed name in v25.1

File: src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al [hunk 1]
```diff
@@ -150,7 +150,7 @@ codeunit 324 "No. Series Copilot Impl."
         Telemetry: Codeunit Telemetry;
         ToolsSelectionPrompt: Text;
     begin
-        if not AzureKeyVault.GetAzureKeyVaultSecret('NoSeriesCopilotToolsSelectionPrompt', ToolsSelectionPrompt) then begin
+        if not AzureKeyVault.GetAzureKeyVaultSecret('NoSeriesCopilotToolsSelectionPromptV2', ToolsSelectionPrompt) then begin // TODO: As the prompt should be different for v25.0 and v25.1, we need to add a new secret for v25.1 and update the code to get the correct prompt based on the version
```
Comment (JesperSchulz) [inline src/Business Foundation/App/NoSeriesCopilot/src/Copilot/NoSeriesCopilotImpl.Codeunit.al]: @darjoo, other than replacing the secret with a new one, do we have a way of versioning prompts?

File: src/Business Foundation/App/NoSeriesCopilot/src/Copilot/Tools/NoSeriesCopNxtYrIntent.Codeunit.al [hunk 2]
```diff
@@ -45,7 +45,7 @@ codeunit 349 "No. Series Cop. Nxt Yr. Intent" implements "AOAI Function"
     var
         AzureKeyVault: Codeunit "Azure Key Vault";
     begin
-        if not AzureKeyVault.GetAzureKeyVaultSecret('NoSeriesCopilotTool3Definition', Definition) then begin
+        if not AzureKeyVault.GetAzureKeyVaultSecret('NoSeriesCopilotTool3DefinitionV2', Definition) then begin // TODO: As the prompt should be different for v25.0 and v25.1, we need to add a new secret for v25.1 and update the code to get the correct prompt based on the version
```
Comment (JesperSchulz) [inline src/Business Foundation/App/NoSeriesCopilot/src/Copilot/Tools/NoSeriesCopNxtYrIntent.Codeunit.al]: One more todo we must deal with before merging the code.

**Review by JesperSchulz — COMMENTED** [review]
Other than the pending TODOs, I'm fine with this PR. But let's wait for Darricks return to see if we cannot find a more elegant solution of versioning prompts, other than creating new ones which replace the old ones.

**Comment by JesperSchulz** [comment]
@DmitryKatson, we should get this ready. Could you ensure the build passes?

**Comment by DmitryKatson** [comment]
@JesperSchulz @darjoo fixed previous build errors

**Comment by DmitryKatson** [comment]
@darjoo just noticed - the learn more link is still not updated.

https://github.com/microsoft/BCApps/blob/main/src/Business%20Foundation/App/NoSeriesCopilot/src/Register/NoSeriesCopilotRegister.Codeunit.al

**Review by JesperSchulz — APPROVED** [review]

**Review by darjoo — APPROVED** [review]

**Pull request merged**